File: src/ai/graph.ts
import { StateGraph, END } from "@langchain/langgraph";
import { ToolNode } from "@langchain/langgraph/prebuilt";
import { AgentAnnotation } from "@/ai/state";
import { agent, routeAgent } from "@/ai/agent";
import { retrieve } from "@/ai/rag";
import { allTools } from "@/ai/tools";

const toolNode = new ToolNode(allTools);

export const graph = new StateGraph(AgentAnnotation)
  .addNode("agent", agent)
  .addNode("retrieve", retrieve)
  .addNode("tools", toolNode)
  .addEdge("__start__", "agent")
  .addConditionalEdges("agent", routeAgent, {
    retrieve: "retrieve",
    tools: "tools",
    [END]: END,
  })
  .addEdge("retrieve", "agent")
  .addEdge("tools", "agent")
  .compile();

File: backend/app/ai/graph.py
from langgraph.graph import END, StateGraph
from langgraph.prebuilt import ToolNode

from app.ai.agent import agent, route_agent
from app.ai.rag import retrieve
from app.ai.state import AgentState
from app.ai.tools import all_tools

graph = (
    StateGraph(AgentState)
    .add_node("agent", agent)
    .add_node("retrieve", retrieve)
    .add_node("tools", ToolNode(all_tools))
    .add_edge("__start__", "agent")
    .add_conditional_edges("agent", route_agent, {"retrieve": "retrieve", "tools": "tools", END: END})
    .add_edge("retrieve", "agent")
    .add_edge("tools", "agent")
    .compile()
)
